qlty-cloud/src/token/auth_flow.rs

```diff
@@ -0,0 +1,234 @@
+use super::DEFAULT_USER;
+use crate::token::{Token, SERVICE};
+use actix_web::{
+    dev::{Server, ServerHandle},
+    get,
+    http::header::LOCATION,
+    rt,
+    web::{Data, Query},
+    App, HttpResponse, HttpResponseBuilder, HttpServer,
+};
+use anyhow::{Ok, Result};
+use console::style;
+use serde::Deserialize;
+use std::{
+    net::TcpListener,
+    sync::{Arc, Mutex},
+    thread,
+    time::Duration,
+};
+use tracing::{debug, info, warn};
+use uuid::Uuid;
+
+const LOGIN_URL: &str = "https://qlty.sh/login";
+const LOGIN_URL_ENV: &str = "QLTY_AUTH_LOGIN_URL";
+
+#[derive(Deserialize, Debug)]
+struct AuthFlowQueryParams {
+    state: String,
+    code: String,
+    redirect_uri: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct AppState {
+    original_state: String,
+    server_handle: Arc<Mutex<Option<ServerHandle>>>,
+    credential_user: String,
+}
+
+impl Default for AppState {
+    fn default() -> Self {
+        AppState {
+            original_state: Uuid::new_v4().to_string(),
+            server_handle: Arc::new(Mutex::new(None)),
+            credential_user: DEFAULT_USER.to_string(),
+        }
+    }
+}
+
+#[get("/")]
+async fn auth_flow(
```

Comment:
This is one synchronous Rust HTTP server we can use to avoid pulling in Tokio and Actix:

https://github.com/tomaka/rouille

Reply:
I think there are some good benefits to relying on the better supported axum/actix libs, one of which is better type-safe support of query string deserialization, something this handler relies heavily on and something that rouille does not seem to support in [its router](https://docs.rs/rouille/latest/rouille/macro.router.html). Testing is also another specific ergonomic benefit, as rouille doesn't seem to have great support there. Finally, rouille doesn't seem to have a very stable approach to stopping the server, something we absolutely require. IMO all of this justifies the small cost of having one or two async uses here.

Reply:
Got it. So it sounds like you appreciate that Actix+Tokio has additional functionality which we can benefit from. Also, it has more users and is more actively developed than Rouille. Based on these factors and others, Actix has better developer ergonomics

I agree those seem like reasonable advantages of Actix+Tokio. They are somewhat the flip side of Actix vs. Rouille in that Rouille is less active, less features, fewer users.

As far as which way to go, I would like to collect more data before making a decision. Would you be up for doing an alternative implementation in Rouille so that we can compare the two?

In case it is helpful, I did a bit of research and found some capabilities that might be useful to paid with a Rouille experiment:
- serde has a typed query string deserializer in that ecosystem — https://docs.rs/serde-querystring/latest/serde_querystring/
- Rouille has a built-in mechanism for constructing fake Requests for testing — https://docs.rs/rouille/latest/rouille/struct.Request.html#method.fake_http
- It looks like Rouille has a mechanism to be shutdown gracefully using a channel communication — https://docs.rs/rouille/latest/rouille/struct.Server.html#method.stoppable